### ISSUE TITLE
Godot 4.3 system resource bug

### ISSUE THREAD

**Issue by Epsteri** [godotengine/godot#96598]
### Tested versions

4.3.stable.steam[77dcf97d8]

### System information

Windows 10, intel core i7-12700f, version 4.3.stable.steam[77dcf97d8]

### Issue description

Godot engine is using too much ram for no reason, even in a blank project it can use up to 8gb of ram sitting at idle.

### Steps to reproduce

let the engine run, and the ram usage will slowly increase.

### Minimal reproduction project (MRP)

[blank_project.zip](https://github.com/user-attachments/files/16886937/blank_project.zip)

**Comment by ImmortalOctogen** [comment]
I completely agree. This is noticeable on my brother's computer with 16 Gb RAM in the game Collectibles! In a few hours, consumption increased from 200 Mb to **3 Gb**. Looks like, this problem starts from 4.2 (I decompiled that game and saw the engine version there). Memory leak?!

**Comment by Epsteri** [comment]
update: I figured out how replication works, you need to use forward+, then delete ui nodes, save, and keep repeating the process, this will pile up ram usage overtime

**Comment by SamDevelopsCode** [comment]
My system: 
OS: Linux Mint 22.1 x86_64 
Kernel: 6.8.0-57-generic 
DE: Cinnamon 6.4.8 
CPU: Intel i5-10500H (12) @ 4.500GHz 
GPU: NVIDIA GeForce RTX 3060 Mobile  
GPU: Intel CometLake-H GT2 [UHD Grap 


Tested on all 3 renderers. Adding in nodes and deleting them is not freeing them from memory.

**Comment by ztc0611** [comment]
We were talking about this in the Discord and I tried to reproduce on Mac. There might be more than one issue, but one of them is definitely that making nodes then deleting them within the editor's `Scene` panel doesn't properly free the RAM. This seems to become an issue much faster on Forward+, but occurs on both Forward+ and Mobile.

Mac:
M4 Macbook Air, 10 Core CPU/10 Core GPU
16GB RAM
macOS 15.4 Sequoia

Reproduction Steps:

-4.4.1.stable
-Blank Project, Forward+
-Sprite2D with icon.svg as its sprite
-⌘+D until there are a couple thousand
-delete them all
-ram never frees

When duplicating the sprites, they were all siblings, there wasn't a parent/child relationship besides with the root node. I deleted them by Shift+Click selecting the entire list then deleting with ⌘+Delete.

In my case at boot the editor (Forward+) was reporting about 1.55GB, and after duplicating 6000 it was 2.15GB. After deleting, it never went below 2.15GB. Also, the size of the image attached to the sprite had no effect. I tried using [this](https://upload.wikimedia.org/wikipedia/commons/5/56/Seal_Rocks%2C_Ocean_Beach%2C_San_Francisco.jpg) photo instead of icon.svg, and there was no difference in the speed at which the ram piled up.

**Comment by jitspoe** [comment]
Even with 32 gigs, leaving Godot running long enough can actually crash a bunch of stuff on Windows due to failing to allocate memory.  The problem for me seems to be the amount of virtual memory reserved for Godot, which can be added as a column in Task Manager called "Commit Size".

<img width="1144" height="86" alt="Image" src="https://github.com/user-attachments/assets/7b531665-5e89-48d2-a3d5-c789cbe707ee" />

Note how even though one instance of Godot is only "using" about 300 megs.  It has a 40 GIGABYTE (!!!) commit size.  This continues to climb until the OS becomes unstable.

Might be unrelated, but I'm kind of wondering if freed stuff doesn't actually get freed, since we can query the pointers to things and see if they're valid, so that memory still has to point somewhere, no?  Or does the pointer get swapped around under the hood?

Edit: Not sure if this is helpful or not, but the most problematic case is my Twitch bot/chat viewer which I sometimes leave running for days.  I think around 3 nodes are added per line of chat:

<img width="1350" height="531" alt="Image" src="https://github.com/user-attachments/assets/26096081-06db-4f71-b42c-f99aa80f0647" />

Chat window, for context:

<img width="455" height="520" alt="Image" src="https://github.com/user-attachments/assets/1c05942c-fd2e-4af9-ba00-22e2f2eb68f1" />

This was the one using 40 gigs of reserved memory.

**Comment by jitspoe** [comment]
Any tips for how to debug this?  It might be something Windows-specific with how the committed memory is allocated.  I'm getting crashes very frequently because of the extreme memory leaks if I leave the editor open for more than a day.

**Comment by Epsteri** [comment]
> Any tips for how to debug this?  It might be something Windows-specific with how the committed memory is allocated.  I'm getting crashes very frequently because of the extreme memory leaks if I leave the editor open for more than a day.

I've noticed that the memory leak happens on the forward+ renderer, switching to compatibility alleviates the issue

**Comment by jitspoe** [comment]
Just for a little more info: I created a blank project in 4.5 stable and just left it open for a few days.  The commit size started at 1,422,454K (which seems pretty crazy just by itself -- why does it need over a gig of memory for just the editor and no assets?) and slowly climbed to 1,671,060K over 1 week.  The active private working set of memory is only 62,692 (with a peak of 1,110,536), so it's not "actively" using that memory, but I guess the way Windows handles memory allocations, the freed memory is never truly freed.

So, while the commit leak is still there on a blank project, it's not nearly as severe.

**Comment by hakuhan** [comment]
I also get memory leak issue. Following pickture is captured by tracy when I just keep character idling in game (Default profiler cannot track this problem):

<img width="1627" height="1051" alt="Image" src="https://github.com/user-attachments/assets/56039317-2411-4100-bb78-c98cf80ac41c" />

<img width="2329" height="187" alt="Image" src="https://github.com/user-attachments/assets/5abb3656-0096-4642-b897-e97512651678" />
